be1-go/hub/base_hub.go

```diff
@@ -0,0 +1,55 @@
+package hub
+
+import (
+	"log"
+	"sync"
+
+	"student20_pop/message"
+
+	"go.dedis.ch/kyber/v3"
+)
+
+type baseHub struct {
+	messageChan chan IncomingMessage
+
+	sync.RWMutex
+	channelByID map[string]Channel
+
+	public kyber.Point
+}
+
+// NewBaseHub returns a Base Hub.
+func NewBaseHub(public kyber.Point) *baseHub {
+	return &baseHub{
+		messageChan: make(chan IncomingMessage),
+		channelByID: make(map[string]Channel),
+		public:      public,
+	}
+}
+
+// RemoveClient removes the client from this hub.
+func (h *baseHub) RemoveClientSocket(client *ClientSocket) {
+	h.RLock()
+	defer h.RUnlock()
+
+	for _, channel := range h.channelByID {
+		channel.Unsubscribe(client, message.Unsubscribe{})
+	}
+}
+
+// Recv accepts a message and enques it for processing in the hub.
+func (h *baseHub) Recv(msg IncomingMessage) {
+	log.Printf("Hub::Recv")
+	h.messageChan <- msg
+}
+
+func start(h Hub, done chan struct{}, messageChan chan IncomingMessage) {
```

Comment:
I tried to create a function like this `func start(h *baseHub, ...` but I encountered a problem: I couldn't find a way to call the correct method `handleIncomingMessage`. This method is different for a witness and an organizer. This is because the `baseHub` is inside the `organizerHub struct` and `witnessHub struct`.  So the function has a interface Hub `h` as an argument so I couldn't do `h.messageChan`. This is why I passed `messageChan` as an argument.

Reply:
But I'm not sure if this is a good idea how I did that.

Reply:
I think it's easier to follow this if baseHub implement `Start(done chan struct{})` with the organizer's implementation in it (it looks like the more generic way of handling a message) and `witnessHub` can override it with its custom implementation.

That way we'd
* Have baseHub implement the Hub interface
* Not require handleIncomingMessage() on the Hub interface

Reply:
Yes I agree! I made the changes.